core/iwasm/common/wasm_c_api.c

```diff
@@ -1651,39 +1651,23 @@ wasm_trap_new_internal(WASMModuleInstanceCommon *inst_comm_rt,
         return NULL;
     }
 
-#if WASM_ENABLE_INTERP != 0
-    if (inst_comm_rt->module_type == Wasm_Module_Bytecode) {
-        error_info = wasm_get_exception((WASMModuleInstance *)inst_comm_rt);
-    }
-#endif
-
-#if WASM_ENABLE_AOT != 0
-    if (inst_comm_rt->module_type == Wasm_Module_AoT) {
-        error_info = aot_get_exception((AOTModuleInstance *)inst_comm_rt);
-    }
-#endif
-
-    /*
-     * a wrong combination of module filetype and compilation flags
-     * also leads to below branch
-     */
-    if (!error_info && !(error_info = default_error_info)) {
-        return NULL;
-    }
-
     if (!(trap = malloc_internal(sizeof(wasm_trap_t)))) {
         return NULL;
     }
 
-    if (!(trap->message = malloc_internal(sizeof(wasm_byte_vec_t)))) {
-        goto failed;
-    }
+    /* fill in message */
+    if (strlen(error_info) > 0) {
+        if (!(trap->message = malloc_internal(sizeof(wasm_byte_vec_t)))) {
+            goto failed;
+        }
 
-    wasm_name_new_from_string_nt(trap->message, error_info);
-    if (strlen(error_info) && !trap->message->data) {
-        goto failed;
+        wasm_name_new_from_string_nt(trap->message, error_info);
+        if (!trap->message->data) {
+            goto failed;
+        }
     }
 
+    /* fill in frames */
```

Comment:
Can also consider merging L1671 to L683, currently the offset of field `frames` in WASMModuleInstance and AOTModuleInstance is the same.